### PR TITLE
Add support for git_revparse

### PR DESCRIPTION
**Pull request by xoofx**
Hi there!
I have added a basic support to [git_revparse](https://libgit2.github.com/libgit2/#HEAD/group/revparse/git_revparse)
As I was not sure how this kind of API is exposed, I have added a static method on the new `RevSpec` instead of polluting the `Repository` class.
Let me know if this is ok.
Thanks!

### CODE REVIEW

**Comment by xoofx** [comment]
Not sure about the errors the ci complains about... looks like a crash of mono...

File: LibGit2Sharp/RevSpec.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                    revSpec.From = GitObject.BuildFrom(repo, Proxy.git_object_id(sh), objType, null);
+                }
+            }
+
+            if (gitRevSpec.To != IntPtr.Zero)
+            {
+                using (var sh = new GitObjectSafeHandle(gitRevSpec.To))
+                {
+                    var objType = Proxy.git_object_type(sh);
+                    revSpec.To = GitObject.BuildFrom(repo, Proxy.git_object_id(sh), objType, null);
```
Comment (dahlbyk) [inline LibGit2Sharp/RevSpec.cs]: Could extract method to DRY this up:

```csharp
var revSpec = new RevSpec
{
    From = BuildObject(gitRevSpec.From),
    To = BuildObject(gitRevSpec.To),
    Type = gitRevSpec.Type,
};
```

**Comment by dahlbyk** [comment]
This would be the only instance of a `public static` dependency on a `Repository` - I think I'd prefer to have `IRepository.RevParse(string)`?

Beyond that, we tend to avoid abbreviations, so maybe:

```csharp
interface IRepository {
    GitObject ParseRevision(string);
    RevisionRangeSpecification ParseRevisionRange(string);
}
```

Also, add test(s) for `RevSpecType.MergeBase`?

File: LibGit2Sharp/LibGit2Sharp.csproj [hunk 2]
```diff
@@ -27,6 +27,7 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <CodeAnalysisRuleSet>AllRules.ruleset</CodeAnalysisRuleSet>
     <DocumentationFile>bin\Debug\LibGit2Sharp.xml</DocumentationFile>
+    <Prefer32Bit>false</Prefer32Bit>
```
Comment (nulltoken) [inline LibGit2Sharp/LibGit2Sharp.csproj]: Could you please elaborate on this change and the one below?

Comment (xoofx) [inline LibGit2Sharp/LibGit2Sharp.csproj]: I don't know why, but VS added it automatically while saving the project. Actually, looking at it, you can't modify it from the VS project itself for this particular .NET4 platform. I can remove it manually though...

File: LibGit2Sharp/RevSpec.cs (hunk 1 above)
Comment (xoofx) [inline LibGit2Sharp/RevSpec.cs]: yes

**Comment by xoofx** [comment]
> Beyond that, we tend to avoid abbreviations, so maybe:

Hm, In fact, I based my code on the fact that there is already a [`RefSpec`](https://github.com/libgit2/libgit2sharp/blob/vNext/LibGit2Sharp/RefSpec.cs) class which was actually not renamed to `ReferenceSpecification`... so not sure which to follow...

So adding `RevParse` methods to `IRepository` directly would be fine?

**Comment by nulltoken** [comment]
@xoofx Thanks a lot for this nice piece of work. However, as far as I can tell, using the current API

 - retrieving a single revparse can be achieved through `IRepository.Lookup()`
 - retrieving a range can be be achieved through `IRepository.Commits.QueryBy()` and providing boundaries via a `CommitFilter`
 - retrieving a merge base can be achieved through `ObjectDatabase.FindMergeBase()`

In order to keep the API terse and more maintainable, we try very hard to not expose similar functionalities through different API endpoints.

Do the already implemented endpoints lack some features that your additions cover? If that's the case, maybe could we rather work in the direction of extending the current endpoints.

**Comment by xoofx** [comment]
@nulltoken I'm actually using this API to validate revspec ([here](https://github.com/xoofx/GitRocketFilter/blob/master/src/RocketFilterApp.cs#L201)), so I believe this API is still relevant, as a Lookup doesn't parse a revspec range.

As a general remark about LibGit2Sharp, while the project is great, I have been frustrated quite a few times with some native APIs that were not fully exposed or cached under lots of abstractions, and I have been forced sometimes to forked it just to access the underlying API (sometimes just to make accessible a call to a native DllImport). (Not specifically for git-rockect-filter, but for other internal projects around one year ago).

Hiding too much of the versatility of the native API, by applying a filter on the API based on an "expected usage" can be quite frustrating for users wanting to take a full leverage on the native API...

Though I'm an occasional user of LibGit2Sharp, so I may completely miss some methods in the framework that could perfectly make the fit... but my main source of documentation and possible usage is coming from libgit2, so I tend to look at what native API I need from there, and then backtrack how they are exposed in LibGit2Sharp...

**Comment by nulltoken** [comment]
@xoofx Thanks for your feedback.

I thought a little more about your proposal. How would you feel about transforming your PR into a simpler static `CommitFilter.BuildFrom()` method which would return an instance of a `CommitFilter`?

**Comment by dahlbyk** [comment]
> How would you feel about transforming your PR into a simpler static `CommitFilter.BuildFrom()` method which would return an instance of a `CommitFilter`?

He has a good point that `CommitFilter.ExcludeReachableFrom` (formerly `Until`) is exclusive, which is not what a revspec range represents.

**Comment by nulltoken** [comment]
>> How would you feel about transforming your PR into a simpler static `CommitFilter.BuildFrom()` method which would return an instance of a `CommitFilter`?

>He has a good point that `CommitFilter.ExcludeReachableFrom` (formerly `Until`) is exclusive, which is not what a revspec range represents.

@dahlbyk I may have read the doc wrong, then

`<rev1>..<rev2>`: Include commits that are reachable from <rev2> but exclude those that are reachable from <rev1>. When either <rev1> or <rev2> is omitted, it defaults to HEAD.

`<rev1>...<rev2>`: Include commits that are reachable from either <rev1> or <rev2> but exclude those that are reachable from both. When either <rev1> or <rev2> is omitted, it defaults to HEAD.

I was thinking of representing them as follows:
 - `<rev1>..<rev2>` -> Include = rev2 / Exclude = rev1
 - `<rev1>...<rev2>` -> Include = rev1;rev2 / Exclude = MergeBase(rev1, rev2)

What did I miss?

**Comment by dahlbyk** [comment]
I've gotten myself all turned around. You are correct on both counts.

Counter-proposal:

* `CommitFilter CommitFilter.Parse(revspec)` (throws on error)
* `bool CommitFilter.TryParse(revspec, out CommitFilter)`

**Comment by carlosmn** [comment]
I'd go with @dahlbyk 's proposal. Making you write `Include = rev1;rev2 / Exclude = MergeBase(rev1, rev2)` makes it very tedious and makes the syntax harder to find. Git has the extended sha-1 syntax in order to avoid having to do this kind of thing.

**Comment by dahlbyk** [comment]
To be clear, I'm pretty sure @nulltoken was suggesting a method to parse the revspec strings into correctly-initialized `CommitFilter`s. I just don't like `BuildFrom()`. :smiling_imp: 

Scope creep: implement `CommitFilter.ToString()` such that it returns the provided `string` when parsing, otherwise building a reasonable revspec.

**Comment by nulltoken** [comment]
>To be clear, I'm pretty sure @nulltoken was suggesting a method to parse the revspec strings into correctly-initialized `CommitFilter`s. I just don't like `BuildFrom()`. :smiling_imp: 

You're right. I'm also notoriously known to be bad at naming things. That doesn't help. :wink:

>Scope creep: implement `CommitFilter.ToString()` such that it returns the provided `string` when parsing, otherwise building a reasonable revspec.

Interesting idea... But that would only deal with the range, not the additional properties (FirstParent, SortStrategy, ....).

**Comment by dahlbyk** [comment]
> > Scope creep: implement CommitFilter.ToString() such that it returns the provided string when parsing, otherwise building a reasonable revspec.
>
> Interesting idea... But that would only deal with the range, not the additional properties (FirstParent, SortStrategy, ....).

Those don't really have a canonical string representation, so I'd be fine excluding them. I'm really not sure why you'd need to `ToString()` on a `CommitFilter` anyway, but it does strike me as an interesting problem. And I love solving problems that don't exist.

**Comment by nulltoken** [comment]
@dahlbyk I've thought a bit more about this. 

Although that seems completely possible to always generate a `CommitFilter` from a string. The opposite isn't true, sadly.

For instance, given a `CommitFilter` with more than one committish in either `Include` or `Exclude`, I haven't found a proper way to express a range.

**Comment by dahlbyk** [comment]
I'm pretty sure this:

```csharp
Include = new[] { "foo", "bar" },
Exclude = new [] { "baz", "quux" },
```

translates to `"foo bar ^baz ^quux"`.

**Comment by nulltoken** [comment]
@dahlbyk Oh! You're right. I thought you were trying to always express the output with "double dot" shortcut. I didn't consider this option.

This should indeed work. :+1:

**Comment by carlosmn** [comment]
That's one possible way that might be written, but there's a few more, and we may not match what was passed in. Are you thinking of using this for the debugger?

**Comment by dahlbyk** [comment]
> That's one possible way that might be written, but there's a few more, and we may not match what was passed in. Are you thinking of using this for the debugger?

The thought would be to capture the parsed string, if provided, otherwise make an educated guess based on specified `Include`/`Exclude`.

```csharp
Assert.Equal("foo..bar", CommitFilter.Parse("foo..bar").ToString());
Assert.Equal("foo...bar", CommitFilter.Parse("foo...bar").ToString());
Assert.Equal("foo..bar", new CommitFilter { Include = "foo", Exclude = "bar" }.ToString());
Assert.Equal("foo bar ^baz", new CommitFilter { Include = new[] { "foo", "bar" }, Exclude = "baz" }.ToString());
```

> Are you thinking of using this for the debugger?

I guess? Not sure, honestly, what the use case would be beyond that.

**Comment by nulltoken** [comment]
So, I've started to work on this `Parse` approach. See https://github.com/libgit2/libgit2sharp/compare/ntk/parse

**Pull request closed without merging**